plugins/http_plugin/beast_http_session.hpp

```diff
@@ -0,0 +1,576 @@
+            }           
+
+            // Access the derived class, this is part of
+            // the Curiously Recurring Template Pattern idiom.
+            Derived& derived()
+            {
+                return static_cast<Derived&>(*this);
+            }
+            
+        public: 
+            // shared_from_this() requires default constructor
+            beast_http_session() = default;
+
+            beast_http_session(
+                std::shared_ptr<http_plugin_state> plugin_state,  
+                std::shared_ptr<eosio::chain::named_thread_pool> thread_pool)
+                : plugin_state_(std::move(plugin_state))
+                , thread_pool_(std::move(thread_pool))
+            {  
+                plugin_state_->requests_in_flight += 1;
+                req_parser_.emplace();
+                req_parser_->body_limit(plugin_state_->max_body_size);
+                res_.emplace();
+
+                session_begin_ = steady_clock::now();
+                read_time_us_ = handle_time_us_ = write_time_us_ = 0;
+
+                // default to true
+                is_send_exception_response_ = true;
+            }
+
+            virtual ~beast_http_session() {
+                is_send_exception_response_ = false;
+                plugin_state_->requests_in_flight -= 1;
+                auto session_time = steady_clock::now() - session_begin_;
+                auto session_time_us = std::chrono::duration_cast<std::chrono::microseconds>(session_time).count();
+                fc_dlog(logger, "session time    ${t}", ("t", session_time_us));                            
+                fc_dlog(logger, "        read    ${t}", ("t", read_time_us_));
+                fc_dlog(logger, "        handle  ${t}", ("t", handle_time_us_));                                            
+                fc_dlog(logger, "        write   ${t}", ("t", write_time_us_));                            
+            }    
+
+            void do_read()
+            {
+                read_begin_ = steady_clock::now();
+
+                // Read a request
+                auto self = derived().shared_from_this();
+                http::async_read(
+                    derived().stream(),
+                    buffer_,
+                    *req_parser_,
+                    [self](beast::error_code ec, std::size_t bytes_transferred) { 
+                        self->on_read(ec, bytes_transferred);
+                    }
+                );                
+            }
+
+            void on_read(beast::error_code ec,
+                         std::size_t bytes_transferred)
+            {
+                boost::ignore_unused(bytes_transferred);
+
+                // This means they closed the connection
+                if(ec == http::error::end_of_stream)
+                    return derived().do_eof();
+
+                if(ec) {
+                    return fail(ec, "read");
+                }
+
+                auto req = req_parser_->get();
```

Comment:
Can call `release()` instead of `get()` to move parsed message out of `req_parser_`